### PR TITLE
JAMES-2630 Migrate CassandraAsyncExecutor.executeReturnApplied consumers to Reactor

### PR DESCRIPTION
**Pull request by blackheaven**

### CODE REVIEW

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLMapper.java [hunk 1]
```diff
@@ -191,16 +191,16 @@ private ACLDiff updateAcl(CassandraId cassandraId, Function<ACLWithVersion, ACLW
...
+                .map(optional -> optional.filter(b -> b).map(any -> acl))
+                .block();
         } catch (JsonProcessingException exception) {
             throw new RuntimeException(exception);
         }
     }
 
     private Optional<ACLWithVersion> getAclWithVersion(CassandraId cassandraId) {
-        ResultSet resultSet = getStoredACLRow(cassandraId).join();
+        ResultSet resultSet = getStoredACLRow(cassandraId).block();
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLMapper.java]: Does this means that "`we block every time we read an ACL`" ?

Can't it be the source of slooooow mailboxes reads?

Can we try to get read of that 'block' here?

This should be fairly easy...

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraModSeqProvider.java [hunk 2]
```diff
@@ -125,81 +114,76 @@ private PreparedStatement prepareSelect(Session session) {
...
-        return runner.executeAsyncAndRetrieveObject(
-            () -> findHighestModSeq(mailboxId)
-                .thenCompose(newModSeq -> tryUpdateModSeq(mailboxId, newModSeq)));
+            }).switchIfEmpty(handleRetries(mailboxId))
+            .map(ModSeq::getValue);
+    }
+
+    private Mono<ModSeq> handleRetries(CassandraId mailboxId) {
+        return findHighestModSeq(mailboxId)
+                .flatMap(newModSeq -> tryUpdateModSeq(mailboxId, newModSeq));
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraModSeqProvider.java]: Euh does that means I can wait indefinitly without never hitting a max retry?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraUidProvider.java [hunk 3]
```diff
@@ -101,66 +91,54 @@ public MessageUid nextUid(MailboxSession mailboxSession, Mailbox mailbox) throws
...
-            })
-            .thenCompose(optional -> {
-                if (optional.isPresent()) {
-                    return CompletableFuture.completedFuture(optional);
-                }
-                return runner.executeAsyncAndRetrieveObject(
-                    () -> findHighestUid(cassandraId)
-                        .thenCompose(readUid -> tryUpdateUid(cassandraId, readUid)));
-            });
+    public Mono<MessageUid> nextUid(CassandraId cassandraId) {
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraUidProvider.java]: Why can't I see `handlesRetry method` in this class?

File: server/container/util/pom.xml [hunk 4]
```diff
@@ -116,6 +116,12 @@
             <groupId>org.slf4j</groupId>
             <artifactId>slf4j-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>backport-util-concurrent</groupId>
+            <artifactId>backport-util-concurrent</artifactId>
```
Comment (chibenwa) [inline server/container/util/pom.xml]: Where do you use that lib? What does it do?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraUidProvider.java (hunk 3 above)
Comment (chibenwa) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraUidProvider.java]: Does this means we do not retry UID creation ???

Damn that explain some cassandra test random failures (MailboxManagerStressTest)

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraModSeqProvider.java (hunk 2 above)
Comment (chibenwa) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraModSeqProvider.java]: I mean, does this class still tries things a bounded number of times like before (& where?)

FunctionRunnerWithRetry for sure have a max retry bound...

**Comment by chibenwa** [comment]
test this please

**Comment by blackheaven** [comment]
Given `org.apache.james.gatling.smtp.scenari.NoAuthenticationNoEncryptionScenario`/`org.apache.james.gatling.smtp.scenari.NoAuthenticationNoEncryptionBigBodyScenario` for 30 minutes:

 * Before (master:df5d678cd43319ba7afca9baabcfc27b338b1898)

Normal:

before-small

Big:

before big

 * After

Normal:

after-small

Big:

after-big

Conclusion: A bit of improvement on small mails, more or less stable on big mails

**Comment by chibenwa** [comment]
Can you be testing also JMAP & IMAP scenari?

Because the mailbox only have an indirect impact on the SMTP protocol...

**Comment by mbaechler** [comment]
merged

**Comment by blackheaven** [comment]
Not better for `org.apache.james.gatling.jmap.scenari.JmapAllScenario` (250 clients / 20 minutes):

 * Before:

master

 * After:

pr2630

**Comment by blackheaven** [comment]
It's worse with a distant Cassandra:

 * Before

master

* After

**Pull request closed without merging**